File: Pharmaceutical_Company_Example/03_Behavioral/C++/Iterator.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <string>
#include <unordered_map>

using namespace std;

// Iterator Interface
class ApplicationIterator {
public:
  virtual bool hasNext() const = 0;
  virtual shared_ptr<unordered_map<string, string>> next() = 0;
  virtual ~ApplicationIterator() = default;
};

// Concrete Iterator
class DrugApplicationIterator : public ApplicationIterator {
private:
  vector<shared_ptr<unordered_map<string, string>>> applications;
  size_t position = 0;
  
public:
  DrugApplicationIterator(const vector<shared_ptr<unordered_map<string, string>>>& apps)
    : applications(apps) {}
  
  bool hasNext() const override {
    return position < applications.size();
  }
  
  shared_ptr<unordered_map<string, string>> next() override {
    if (hasNext()) {
      return applications[position++];
    }
    return nullptr;
  }
};

// Aggregate Interface
class ApplicationCollection {
public:
  virtual shared_ptr<ApplicationIterator> createIterator() = 0;
  virtual ~ApplicationCollection() = default;
};

// Concrete Aggregate
class DrugApplicationCollection : public ApplicationCollection {
private:
  vector<shared_ptr<unordered_map<string, string>>> applications;
  
public:
  void addApplication(const unordered_map<string, string>& application) {
    applications.push_back(make_shared<unordered_map<string, string>>(application));
  }
  
  shared_ptr<ApplicationIterator> createIterator() override {
    return make_shared<DrugApplicationIterator>(applications);
  }
};

int main() {
  DrugApplicationCollection drugCollection;
  
  // Adding applications
  drugCollection.addApplication({{"drug_name", "Test Vaccine 01"}, {"status", "Under Review"}});
  drugCollection.addApplication({{"drug_name", "Test Small Molecule 01"}, {"status", "Approved"}});
  drugCollection.addApplication({{"drug_name", "Test Monoclonal Antibody 01"}, {"status", "Rejected"}});
  
  // Create iterator
  auto iterator = drugCollection.createIterator();
  
  // Use the iterator to traverse the applications
  while (iterator->hasNext()) {
    auto application = iterator->next();
    if (application) {
      cout << "Drug Name: " << (*application)["drug_name"] << " - Status: " << (*application)["status"] << endl;
    }
  }
  
  return 0;
}

File: Pharmaceutical_Company_Example/03_Behavioral/Java/Iterator.java
import java.util.List;
import java.util.ArrayList;
import java.util.Map;

// Iterator Interface
interface ApplicationIterator {
    boolean hasNext();
    Map<String, String> next();
}

// Concrete Iterator
class DrugApplicationIterator implements ApplicationIterator {
    private List<Map<String, String>> applications;
    private int position;

    public DrugApplicationIterator(List<Map<String, String>> applications) {
        this.applications = applications;
        this.position = 0;
    }

    @Override
    public boolean hasNext() {
        return position < applications.size();
    }

    @Override
    public Map<String, String> next() {
        if (hasNext()) {
            Map<String, String> app = applications.get(position);
            position++;
            return app;
        }
        return null;
    }
}

// Aggregate Interface
interface ApplicationCollection {
    ApplicationIterator createIterator();
}

// Concrete Aggregate
class DrugApplicationCollection implements ApplicationCollection {
    private List<Map<String, String>> applications;

    public DrugApplicationCollection() {
        this.applications = new ArrayList<>();
    }

    public void addApplication(Map<String, String> application) {
        applications.add(application);
    }

    @Override
    public ApplicationIterator createIterator() {
        return new DrugApplicationIterator(applications);
    }
}

public class Main {
    public static void main(String[] args) {
        DrugApplicationCollection drugCollection = new DrugApplicationCollection();
        
        // Adding applications
        drugCollection.addApplication(Map.of("drug_name", "Test Vaccine 01", "status", "Under Review"));
        drugCollection.addApplication(Map.of("drug_name", "Test Small Molecule 01", "status", "Approved"));
        drugCollection.addApplication(Map.of("drug_name", "Test Monoclonal Antibody 01", "status", "Rejected"));
        
        // Create the iterator
        ApplicationIterator iterator = drugCollection.createIterator();
        
        // Use the iterator to traverse the applications
        while (iterator.hasNext()) {
            Map<String, String> application = iterator.next();
            if (application != null) {
                System.out.println("Drug Name: " + application.get("drug_name") + " - Status: " + application.get("status"));
            }
        }
    }
}
